src/cpp/src/visual_language/qwen2vl/classes.cpp

```diff
@@ -1477,4 +1511,516 @@ ov::Tensor InputsEmbedderQwen2VL::create_position_ids(
     return position_ids;
 }
 
+// [CDPruner] Text feature extraction functions implementation
+
+std::vector<int64_t> InputsEmbedderQwen2VL::extract_instruction_tokens(const ov::Tensor& input_ids,
+                                                                       int64_t image_pad_token_id,
+                                                                       int64_t vision_start_token_id,
+                                                                       int64_t vision_end_token_id) const {
+    std::vector<int64_t> instruction_tokens;
+
+    // Get input_ids data
+    const int64_t* input_ids_data = input_ids.data<const int64_t>();
+    size_t seq_len = input_ids.get_shape()[1];  // [batch_size, seq_len]
+
+    bool inside_vision_region = false;
+
+    // Iterate through the sequence to extract instruction tokens
+    for (size_t i = 0; i < seq_len; ++i) {
+        int64_t current_token = input_ids_data[i];
+
+        // Check for vision start token
+        if (current_token == vision_start_token_id) {
+            inside_vision_region = true;
+            continue;
+        }
+
+        // Check for vision end token
+        if (current_token == vision_end_token_id) {
+            inside_vision_region = false;
+            continue;
+        }
+
+        // Skip if inside vision region or if it's an image pad token
+        if (inside_vision_region || current_token == image_pad_token_id) {
+            continue;
+        }
+
+        // This is a valid instruction token
+        instruction_tokens.push_back(current_token);
+    }
+
+    return instruction_tokens;
+}
+
+// [CDPruner] Override: Text feature extraction for relevance calculation
+ov::Tensor InputsEmbedderQwen2VL::extract_text_features_for_pruning(const ov::Tensor& text_embeds,
+                                                                    const ov::Tensor& input_ids,
+                                                                    int64_t vision_start_token_id,
+                                                                    int64_t vision_end_token_id) const {
+    // Qwen2VL uses image_pad_token_id which we need to get from vision_token_ids
+    int64_t image_pad_token_id = m_vision_token_ids.at("image_pad");
+
+    // Extract instruction tokens
+    std::vector<int64_t> instruction_tokens =
+        extract_instruction_tokens(input_ids, image_pad_token_id, vision_start_token_id, vision_end_token_id);
+
+    if (instruction_tokens.empty()) {
```

Comment:
Does it mean that CDPruner focuses only on image diversity if there is no text prompt?